pkg/keycloak/client.go

```diff
@@ -678,25 +403,27 @@ func defaultRequester() Requester {
 }
 
 type KeycloakInterface interface {
-	ListRealms() ([]*v1alpha1.KeycloakRealm, error)
-	GetRealm(realmName string) (*v1alpha1.KeycloakRealm, error)
 	CreateRealm(realm *v1alpha1.KeycloakRealm) error
-	UpdateRealm(realm *v1alpha1.KeycloakRealm) error
+	GetRealm(realmName string) (*v1alpha1.KeycloakRealm, error)
+	UpdateRealm(kcRealm, specRealm *v1alpha1.KeycloakRealm) error
+	DeleteRealm(realmName string) error
+	ListRealms() ([]*v1alpha1.KeycloakRealm, error)
 
-	GetClient(clientID, realmName string) (*v1alpha1.KeycloakClient, error)
 	CreateClient(client *v1alpha1.KeycloakClient, realmName string) error
+	GetClient(clientID, realmName string) (*v1alpha1.KeycloakClient, error)
+	UpdateClient(kcClient, specClient *v1alpha1.KeycloakClient, realmName string) error
 	DeleteClient(clientID, realmName string) error
-	UpdateClient(specClient *v1alpha1.KeycloakClient, realmName string) error
 	ListClients(realmName string) ([]*v1alpha1.KeycloakClient, error)
 
 	CreateUser(user *v1alpha1.KeycloakUser, realmName string) error
+	GetUser(userID, realmName string) (*v1alpha1.KeycloakUser, error)
+	UpdateUser(kcUser, specUser *v1alpha1.KeycloakUser, realmName string) error
 	DeleteUser(userID, realmName string) error
-	UpdateUser(specUser *v1alpha1.KeycloakUser, realmName string) error
 	ListUsers(realmName string) ([]*v1alpha1.KeycloakUser, error)
 
 	CreateIdentityProvider(identityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
 	GetIdentityProvider(alias, realmName string) (*v1alpha1.KeycloakIdentityProvider, error)
-	UpdateIdentityProvider(identityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
+	UpdateIdentityProvider(kcIdentityProvider, specIdentityProvider *v1alpha1.KeycloakIdentityProvider, realmName string) error
 	DeleteIdentityProvider(alias, realmName string) error
 	ListIdentityProviders(realmName string) ([]*v1alpha1.KeycloakIdentityProvider, error)
```

Comment:
I don't think the keycloak client needs to be aware of spec and kc objects, we should just pass one object in that is the thing we want to update.

If IDs or whatever really need to be passed in they should be passed explicitly, but we intended to have the spec object have all fields set from the kc object so it should have everything needed to create a request.